FEXCore/Source/Interface/Core/JIT/JIT.cpp

```diff
@@ -516,25 +512,23 @@ static uint64_t Arm64JITCore_ExitFunctionLink(FEXCore::Core::CpuStateFrame* Fram
   }
 
   uintptr_t branch = (uintptr_t)(Record)-8;
+  LOGMAN_THROW_A_FMT((branch % 16) == 0, "Incorrect alignment for block linking record");
 
   auto offset = HostCode / 4 - branch / 4;
   if (ARMEmitter::Emitter::IsInt26(offset)) {
-    // optimal case - can branch directly
-    // patch the code
-    ARMEmitter::Emitter emit((uint8_t*)(branch), 4);
-    emit.b(offset);
+    // This is the optimal case, where the target can be encoded in a single instruction.
+    // Atomically patch the code with a relative branch.
+    const uint32_t Patch = (0b0001'01 << 26) | (offset & ((1u << 26) - 1));
+    std::atomic_ref<uint32_t>(*reinterpret_cast<uint32_t*>(branch)).store(Patch, std::memory_order::relaxed);
     ARMEmitter::Emitter::ClearICache((void*)branch, 4);
-
-    // Add de-linking handler
-    Thread->LookupCache->AddBlockLink(GuestRip, Record, DirectBlockDelinker);
   } else {
     // fallback case - do a soft-er link by patching the pointer
-    Record->HostBranch = HostCode;
-
-    // Add de-linking handler
-    Thread->LookupCache->AddBlockLink(GuestRip, Record, IndirectBlockDelinker);
+    std::atomic_ref<uint64_t>(Record->HostBranch).store(HostCode, std::memory_order::seq_cst);
```

Comment:
Does this need a cache clear, or is all that taken care of by `seq_cst`?

Reply:
That's actually a good question.  I think because the other threads executing are loading data with out atomics then there needs to be a `dc cvau, <line>; dsb ish;` to ensure visibility?

Reply:
Yes, I've also done more research and you're right. We either need a barrier in the reading thread (bad), or a cache clean in the writing thread (good). The other parts should be good, since the more expensive ClearICache [already includes those barriers](https://github.com/llvm/llvm-project/blob/278c429d11e63bc709ea8c537b23c4e350ce2a07/compiler-rt/lib/builtins/clear_cache.c#L125-L157).

btw here's a good reference for cache and barrier instructions on arm: https://mariokartwii.com/armv8/ch30.html . (Certainly impressive that the Mario Kart community write clearer documentation on this than arm themselves.)